[Reviewer: Chris] Re-raises alarms every 30 seconds

AME2 opened the pull request:
Tests both the alarm refactoring (creating a new class called MultiStateAlarms for alarms which can be raised with two or more possible states) and the alarm reraising function (which reraises all alarms every 30 seconds).

AME2 committed:
Re-raises alarms every 30 seconds

rkday-pro commented:
You've put [Reviewer: Rob] but assigned it to Chris - I'll assume Chris is handling unless you assign it to me.

chris-elford-metaswitch commented:
The tests here are somewhat more complicated than those laid out in your test plan. Can you please simplify a little to make this easier to read, and add some comments explaining what your tests aim to prove. Simplicity is your friend.

AME2 committed:
Applies markups to UTs

AME2 committed:
Applies markups to UTs

AME2 commented:
I've applied CNE's markups.

chris-elford-metaswitch commented on src/ut/alarm_test.cpp:
```diff
@@ -162,6 +167,305 @@ MATCHER_P(VoidPointeeEqualsStr, value, "")
   return (strcmp((char*)arg, value) == 0);
 }
 
+// Raise a MultiStateAlarm in two of its possible states and expects two ZMQ
```
*expect

chris-elford-metaswitch commented on src/ut/alarm_test.cpp:
```diff
@@ -162,6 +167,305 @@ MATCHER_P(VoidPointeeEqualsStr, value, "")
+  _multi_state_alarm.set_major();
+  // Causes the test thread to wait to ensure the zmq messages get sent before
+  // the test terminates.
+  for (unsigned int i = 0; i < 2; i++)
```
A for loop seems a bit much!

chris-elford-metaswitch commented on src/ut/alarm_test.cpp:
```diff
@@ -162,6 +167,305 @@ MATCHER_P(VoidPointeeEqualsStr, value, "")
+// Raises an Alarm (single state) and then simulates time moving forward by
+// thirty seconds. We should expect three ZMQ messages: the first caused by the
+// alarm being raised, the second and third caused by the function to re-send
+// alarms every 30 seconds.
```
Naively, I'd expect only one re-raise as we've only waited 30s once. It looks like you re-raise the first-raised alarm straight away in each process using this infrastructure. That's not ideal. Is it simple to fix?

chris-elford-metaswitch replied:
Ah, I think I got that wrong. In fact, it's two messages because there are two alarms. I'd explain that here rather than in the middle of the test.

chris-elford-metaswitch commented on src/ut/alarm_test.cpp:
```diff
@@ -423,7 +727,7 @@ TEST_F(AlarmZmqErrorTest, CloseSocket)
   EXPECT_TRUE(AlarmReqAgent::get_instance().start());
   AlarmReqAgent::get_instance().stop();
-
+  
```
Oops

chris-elford-metaswitch commented:
Not much to add. These tests are much simpler, and seem more targeted at individual pieces of function. I'd have been happy without a couple of them as I don't think they exercise new function, but let's keep them it case that changes.

AME2 committed:
Fixes threading race condition

AME2 committed:
Fixes threading race condition

AME2 commented:
I've expanded the comments to explain why, after moving time forward by thirty seconds, we receive a ZMQ message corresponding to clearing an alarm when we have never seen that alarm be cleared before.

AME2 commented:
I've removed the _alarmed private member function of the class BaseAlarm and have modified the alarmed() public method. This method now looks at the _last_state_raised private member variable to decide whether an alarm is currently active (in a non-CLEARED severity) or not. Then to stop the threading race condition I've included a mutex lock in my set_x methods.
In class BaseAlarm I've also moved the _first_alarm_raised variable into the private namespace and now use the methods start_resending_alarms and has_alarm_been_raised to write and read this variable.

AME2 committed:
Replaces ambiguous variable name

AME2 committed:
Includes changes to cpp-common

AME2 committed:
Explains Alarm Manager assuming unknown alarms are CLEARED

AME2 committed:
Explains the call to call_complete at the end of each test

AME2 committed:
Merge remote-tracking branch 'origin/master' into ReraisingAlarms
